### DISCUSSION TITLE
replace text and keep original color in cmyk

### DISCUSSION THREAD

**Discussion by ricardo85x** [pymupdf/PyMuPDF#1849] [Looking for help] [answered]
I am replacing the text of a pdf

The best solution I came up with was taking the coordinates, fonts and color and then creating a new text with TextWriter.

Visually everything seems fine, but when I went to print the pdf the color of the text that (I changed/created) was completely different from the color of a text that I didn't change.

I'm using this code to apply color to the text:

```python
text_writer = fitz.TextWriter(
            page.rect, 
            color=fitz.sRGB_to_pdf(textbox_info["color"]) # color from the page.get_text("dict")
)
```

I believe the problem is in this sRGB_to_pdf.
I didn't find anything in the documentation about cmyk_to_pdf.

Is there any way to copy the color of an existing text using cmyk instead of RGB?

**Comment by JorjMcKie** [comment]
Interesting problem ...
The "dict" output contains an sRGB integer version of the original color. The conversion from the original color space in the PDF to RGB is done inside MuPDF, and it always is RGB. There is no way of knowing what the original colorspec has been inside PDF.
If you would know the color to use you could output any supported color space, including CMYK.
Are you sure that `sRGB_to_pdf()`itself is wrong?

**Reply by ricardo85x** [reply]
I can't control the color that comes in the pdf text.

I know it's CMYK, and pantone is also used.

Visually it is almost similar, but when printed it is completely different.
The color looks washed out

It must be because the color is converted from RGB.

I'm trying to find out if the original color is saved in the pdf, but I haven't had much luck

```
BT
/CS2 cs 1  scn
/T1_2 1 Tf
74.7583 0 0 74.7583 1571.8821 767.0728 Tm
[(L)-5 (A)25 (S)10 (TNAMER)]TJ
ET
```

**Comment by JorjMcKie** [comment, marked as answer]
What I don't understand however is, how can things be "visually" ok, but not on the printer?

**Reply by ricardo85x** [reply]
Visually it is almost similar, but when printed it is completely different.
The color looks washed out

**Comment by terminalh2t3** [comment]
I've also got the same problem.
RGB normally covers more color range than CMYK. However, CMYK is better for printers due to the limitation of how color printers can work.
In my case, I want to access all drawings on the page `page.get_cdrawings()`, and then verify the colors whether match with the printing requirements or not. But since PyMuPDF converted it all to RGB so there is no way to convert it back.

In Adobe Illustrator, the PDF file show CMYK is: 100, 0, 0, 0. PyMuPDF converted to RGB (0.0, 0.6809338331222534, 0.9379873275756836) => mean (R: 0, G: 174, B: 239). Unfortunately, there is no way to convert it back to 100, 0, 0, 0.

Should we store the original colors instead of convert to RGB? For example, instead of returns RGB values, we should return the ColorSchema object that allow to convert or manipulate?

**Reply by JorjMcKie** [reply]
There is no way to locate the original color specification of text or drawings (CMYK or whatever that was). You have to work with the returned sRGB / RGB plus transparency information.

**Reply by terminalh2t3** [reply]
Way work around I did so far (in my case, PDF exported from Adobe Illustration)
1. Parse raw data from content stream and find all CMYK color used in the documents (handle: fill, stroke, tint colors, colorrants) and convert to RGB (use ImageCms.profileToProfile), saved in local.
2. Mapping the RGB color parsed by MuPDF to the RGB color parse in step 1 (can apply for: get_drawings, get_text)

Just an example:
```python
# Parse CMYK colors from content stream
content_stream = self.get_page_content_stream()

# Parse all line the content stream by regex ^\d[\d|\.| ]+\d (k|K)$
cmyk_colors = {}
for line in content_stream.split("\n"):
    if re.match(r"^\d[\d|\.| ]+\d (k|K)$", line):
        cmyk_key = line.replace(" ", "_")
        if cmyk_key not in cmyk_colors:
            cmyk = [float(v) for v in line.split(" ")[:4]]
            cmyk_colors[cmyk_key] = {
                "cmyk": cmyk, 
                "mode": "CMYK"
            }

...

# Map CMYK from RGB
def map_cmyk_from_rgb(self, r, g, b):
    colors_dict = self.cymk_colors
    nearest_color = None
    current_distance = 1000000 # Set to a large number
    for _, color_info in colors_dict.items():
        # Calculate distance between two RGB colors
        in_rgb_color = [r, g, b]
        query_rgb_color = color_info['rgb_full_scale']
        in_rgb_color = [round(v * 255) for v in in_rgb_color]

        distance = sum([(a - b) ** 2 for a, b in zip(in_rgb_color, query_rgb_color)])
        
        if distance == 0.0:
            return color_info, 0.0
        else:
            if distance < current_distance:
                nearest_color = color_info
                current_distance = distance
    return nearest_color, current_distance
```
